client/X11/xf_cliprdr.c

```diff
@@ -1157,6 +1224,66 @@ static UINT xf_cliprdr_send_client_format_list_response(xfClipboard* clipboard,
 	return clipboard->context->ClientFormatListResponse(clipboard->context, &formatListResponse);
 }
 
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT xf_cliprdr_send_client_file_contents(xfClipboard* clipboard, UINT32 streamId,
+                                                 UINT32 listIndex, UINT32 nPositionLow,
+                                                 UINT32 nPositionHigh, UINT32 cbRequested)
+{
+	CLIPRDR_FILE_CONTENTS_REQUEST formatFileContentsRequest;
+	formatFileContentsRequest.streamId = streamId;
+	formatFileContentsRequest.listIndex = listIndex;
+	formatFileContentsRequest.dwFlags = FILECONTENTS_RANGE;
+	formatFileContentsRequest.nPositionHigh = nPositionHigh;
+	formatFileContentsRequest.nPositionLow = nPositionLow;
+	formatFileContentsRequest.cbRequested = cbRequested;
+	formatFileContentsRequest.haveClipDataId = FALSE;
+	return clipboard->context->ClientFileContentsRequest(clipboard->context,
+	                                                     &formatFileContentsRequest);
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT
+xf_cliprdr_server_file_contents_response(CliprdrClientContext* context,
+                                         const CLIPRDR_FILE_CONTENTS_RESPONSE* fileContentsResponse)
+{
+	UINT32 count;
+	UINT32 index;
+	BOOL found = FALSE;
+	xfCliprdrFuseStream* stream;
+	xfClipboard* clipboard = (xfClipboard*)context->custom;
+	UINT32 stream_id = fileContentsResponse->streamId;
+
+	ArrayList_Lock(clipboard->stream_list);
+	count = ArrayList_Count(clipboard->stream_list);
+
+	for (index = 0; index < count; index++)
+	{
+		stream = (xfCliprdrFuseStream*)ArrayList_GetItem(clipboard->stream_list, index);
+
+		if (stream->stream_id == stream_id)
+		{
+			found = TRUE;
+			break;
+		}
+	}
+	ArrayList_Unlock(clipboard->stream_list);
+	if (found)
+	{
+		fuse_reply_buf(stream->req, (const char*)fileContentsResponse->requestedData,
+		               fileContentsResponse->cbRequested);
+		ArrayList_RemoveAt(clipboard->stream_list, index);
+	}
```

Comment:
I believe the lock needs to be held here until after the found item has been removed from the list.

Reply:
After a double check , i found that  calling `ArrayList_RemoveAt` before `ArrayList_Unlock` will cause deadlock if  ArrayList_New with `BOOL synchronized=TRUE`. :disappointed:

Reply:
You're not wrong in that it will cause a double lock if the collection is synchronized. However, given that almost everywhere you are locking it manually and inspecting it via ArrayList_GetItem(), I think it could be instantiated with `synchronized` set to false. Then ArrayList_RemoveAt() won't add more locks on top of whatever is done manually with ArrayList_Lock().